modin/utils.py

```diff
@@ -11,53 +11,141 @@
 # governing permissions and limitations under the License.
 
+import types
+
 import pandas
 import numpy as np
 
-from modin.config import Engine, Backend, IsExperimental
+from modin.config import Engine, Backend, IsExperimental, DocstringUrlTestMode
 
+PANDAS_API_URL_TEMPLATE = f"https://pandas.pydata.org/pandas-docs/version/{pandas.__version__}/reference/api/{{}}.html"
 
-def _inherit_func_docstring(source_func):
-    """Define `func` docstring from `source_func`."""
+if DocstringUrlTestMode.get():
+    # a list which contains all generated urls so they could be checked for being valid
+    _GENERATED_URLS = set()
 
-    def decorator(func):
-        func.__doc__ = source_func.__doc__
-        return func
 
-    return decorator
+def _replace_doc(
+    source_obj, target_obj, overwrite, apilink, parent_cls=None, attr_name=None
+):
+    """Replaces docstring in `target_obj`, possibly taking from `source_obj` and augmenting.
 
+    Can append the link to Pandas API online documentation.
 
-def _inherit_docstrings(parent, excluded=[]):
-    """Creates a decorator which overwrites a decorated class' __doc__
+    Parameters
+    ----------
+    source_obj : object
+        Any object from which to take docstring from.
+    target_obj : object
+        The object which docstring to replace
+    overwrite : bool
+        Forces replacing the docstring with the one from `source_obj` even
+        if `target_obj` has its own non-empty docstring
+    apilink : str
+        If non-empty, insert the link to Pandas API documentation.
+        Should be the prefix part in the URL template, e.g. "pandas.DataFrame".
+    parent_cls : class, optional
+        If `target_obj` is an attribute of a class, `parent_cls` should be that class.
+        This is used for generating the API URL as well as for handling special cases
+        like `target_obj` being a property.
+    attr_name : str, optional
+        Gives the name to `target_obj` if it's an attribute of `parent_cls`.
+        Needed to handle some special cases and in most cases could be determined automatically.
+    """
+    source_doc = source_obj.__doc__ or ""
+    target_doc = target_obj.__doc__ or ""
+    doc = source_doc if overwrite or not target_doc else target_doc
+
+    if parent_cls and not attr_name:
+        if isinstance(target_obj, property):
+            attr_name = target_obj.fget.__name__
+        elif isinstance(target_obj, (staticmethod, classmethod)):
+            attr_name = target_obj.__func__.__name__
+        else:
+            attr_name = target_obj.__name__
+
+    if (
+        source_doc.strip()
+        and apilink
+        and "Pandas API documentation <" not in target_doc
+        and (not (attr_name or "").startswith("_"))
+    ):
+        if attr_name:
+            token = f"{apilink}.{attr_name}"
+        else:
+            token = apilink
+        url = PANDAS_API_URL_TEMPLATE.format(token)
+        if DocstringUrlTestMode.get():
```

Comment:
Instead of special casing this for the tests, can we factor out the URL generation independent of this and test that independently?

Reply:
I'm not sure I understand your thought.

I could change the test so it would parse URLs out of docstrings (using `re` or smth like this), which would eliminate the need for separate list of URLs.

But I don't like the idea of enumerating every object and its properties...

So I'm not sure what you want to achieve here :)

Reply:
**Gregory** gives me a pointer that what you actually wanted from me here is a separate (unit) test that checks the "add link" functionality (a test for the decorator). This is certainly possible, but it won't replace what I'm testing here - I'm testing that the decorator was used _correctly_ and that the decorator, when used, generated URLs that really _exist_.

Reply:
I see. I prefer to keep special casing testing code out of the implementation. The URL generation test is important, so would it be sufficient to test that part as a unit outside of the logic? We are only testing that the URL is valid, so I believe it should be possible to do.

Reply:
As greed over personal communication, I've removed this testing tidbits and replaced them by a moneky-patchable function for URL generation. I hope it works better for everyone now.